### DISCUSSION TITLE
Execution Plan Output and Params explanation

### DISCUSSION THREAD

**Discussion by ianmark72** [AbsaOSS/spline-spark-agent#563] [Q&A]
I’m looking for some details on certain fields of the execution plan; specifically, how the fields “output” and “params” are generated. I’m working on an app that overrides the default dispatcher to perform a transformation to the spline output. I want to make sure I’m properly translating the element mappings between operations which seems to depend on these fields. Below are some examples of operations with varying values for the keys. Can anyone provide an explanation of how these values are populated? 

```json
{
  "id": "op-6",
  "name": "Project",
  "childIds": [
    "op-7"
  ],
  "output": [
    "attr-6",
    "attr-7",
    "attr-8"
  ],
  "params": {
    "projectList": [
      {
        "__exprId": "expr-3"
      },
      {
        "__exprId": "expr-4"
      },
      {
        "__exprId": "expr-5"
      }
    ]
  }
}
{
  "id": "op-5",
  "name": "Project",
  "childIds": [
    "op-6"
  ],
  "output": [
    "attr-6",
    "attr-7",
    "attr-8",
    "attr-9"
  ],
  "params": {
    "projectList": [
      {
        "__attrId": "attr-6"
      },
      {
        "__attrId": "attr-7"
      },
      {
        "__attrId": "attr-8"
      },
      {
        "__exprId": "expr-6"
      }
    ]
  }
},
{
  "id": "op-1",
  "name": "Intersect",
  "childIds": [
    "op-2",
    "op-8"
  ],
  "output": [
    "attr-6",
    "attr-7",
    "attr-8",
    "attr-9"
  ],
  "params": {
    "isAll": false
  }
}
```

**Comment by cerveada** [comment]
> I’m looking for some details on certain fields of the execution plan; specifically, how the fields “output” and “params” are generated.

Look at `OperationNodeBuilder` trait and it's children. In `GenericPlanNodeBuilder` you can see creation of `DataOperation` including `output` and `params`. But I recommend to look at all of the children of `OperationNodeBuilder`, because different operation types can do it slightly different. 

>  I want to make sure I’m properly translating the element mappings between operations which seems to depend on these fields.

You probably know how the operations are connected via `childIds`. Starting from write operation looking at the `childIds `  the operation with that id and is the previous lineage operation. This way you can walk the whole lineage graph from write to all reads.

There is another graph for attributes, expressions and literals. Following their id's you can get from attributes that are used in the Write operation all the way up to the read attributes.

Connection between these two graphs are the `output` and sometimes `params`. Output is telling you exactly what attributes are the **output** of the operation. Params is for various types of attribute and expression references stored in Spark, depending on each operation type this can be different.

So if you want to keep the lineage graphs valid, you must make sure that your modification will not break any of those connections.
